R/tidy_output.R

```diff
@@ -0,0 +1,318 @@
+    sequence_index = .make_index_column(shape = shape, which = 1L),
+    layer_index = -1L, # just a placeholder
+    head_index = .make_index_column(shape = shape, which = 2L),
+    token_index = .make_index_column(shape = shape, which = 3L),
+    segment_index = -1L, # just a placeholder
+    token = "", # just a placeholder
+    token_id = -1L, # just a placeholder
+    attention_token_index = .make_index_column(shape = shape, which = 4L),
+    attention_segment_index = -1L, # just a placeholder
+    attention_token = "", # just a placeholder
+    attention_token_id = -1L # just a placeholder
+  )
+
+  layer_indexes <- seq_len(length(attention_output))
+
+  sequence_attention <- purrr::map_dfr(
+    layer_indexes,
+    function(this_layer_index) {
+      result_i <- torch::as_array(attention_output[[this_layer_index]])
+
+      this_layer <- torch::torch_flatten(result_i,
+                                         start_dim = 1L,
+                                         end_dim = 4L)
+
+      this_layer <- tibble::tibble(attention_weight = as.numeric(this_layer))
+
+      this_layer <- dplyr::bind_cols(index_columns, this_layer)
+
+      this_layer[["layer_index"]] <- this_layer_index # yay base R
+      this_layer
+    }
+  )
+  return(
+    dplyr::arrange(
+      sequence_attention,
+      .data$sequence_index
+    )
+  )
+}
+
+
+# .finalize_attention -----------------------------------------------------
+
+
+#' Finalize Attention Output
+#'
+#' Takes the data frame output from `.process_attention_result`, joins on the
+#' `token_map` information and filters out the padding tokens.
+#'
+#' @param processed_attention The output from `.process_attention_result`.
+#' @param token_map The output from `.make_token_map`.
+#'
+#' @return A tidy version of the attention weights matrix, with one weight value
+#'   per row.
+#' @keywords internal
+.finalize_attention <- function(processed_attention,
+                                token_map) {
+  to_ret <- processed_attention %>%
+    # first add columns for attending tokens
+    dplyr::left_join(token_map,
+                     by = c("sequence_index", "token_index"),
+                     suffix = c("", "_fill")) %>%
+    dplyr::mutate(segment_index = .data$segment_index_fill,
+                  token = .data$token_fill,
+                  token_id = .data$token_id_fill) %>%
+    dplyr::select(-dplyr::ends_with("_fill")) %>%
+    # now add columns for attendee tokens
+    dplyr::left_join(token_map,
+                     by = c("sequence_index",
+                            "attention_token_index" = "token_index"),
+                     suffix = c("", "_fill")) %>%
+    dplyr::mutate(attention_segment_index = .data$segment_index_fill,
+                  attention_token = .data$token_fill,
+                  attention_token_id = .data$token_id_fill) %>%
+    dplyr::select(-dplyr::ends_with("_fill")) %>%
+    # I don't think we need to support other conventions for the pad token.
```

Comment:
I think this is another thing that points at the idea that we should make sure that {tt} tells us that the output we're working with is valid.